### DISCUSSION TITLE
Wrong user getting displayed on front end

### DISCUSSION THREAD

**Discussion by DrewBrasher** [OrchardCMS/OrchardCore#11911] [Q&A] [answered]
I have an OrchardCore (1.4) site that is hosted on Windows Server 2008 R2 and IIS 7.5. I have created a theme for the front and copied the LoginMenu.cshtml from TheTheme (https://github.com/OrchardCMS/OrchardCore/blob/9c8f94ff18a462a2d7c0f1feaf4397000f636849/src/OrchardCore.Themes/TheTheme/Views/LoginMenu.cshtml) into my theme.

If I log in as one user, then log out and log in as another user sometimes the front end still shows the previous user is logged in but if I go to the admin side it shows the correct user. I can go back to the front end and it still has the incorrect user name displayed. I disable caching in my browser and it still is wrong. If I wait 5 or 10 minutes and refresh the page it will show the correct user.

This also happens sometimes when one person logs in on their computer and then another person logs in on a different computer.

I have not been able to reproduce the issue locally so I'm thinking this is more of a server/IIS issue than OrchardCore but if anyone has any ideas what could cause this I would appreciate it.

**Comment by Skrypt** [comment, marked as answer]
Have you checked the "remember me" option when you are logging in so that it persists a cookie? Because the admin side has its own cookie that uses the tenant ID or name if I remember correctly. You need to define if you are using 2 different cookies or only 1 for both admin and frontend. Right now it looks like it is not using the same cookie to define the current user.

**Reply by DrewBrasher** [reply]
I have not been checking the remember me box.  Here are the cookies I have without it checked:
```
admin_Default (session)
orchauth_Default (session)
Default-adminPreferences (expires in 2 weeks)
```
I tried checking the remember me box and the only difference I see is that the `orchauth_Default` is set to expire in 2 weeks instead of session. I also still have the issue with the box checked.

**Reply by Skrypt** [reply]
Ok, thanks I will do some tests on my side. Though, as I said, it looks like the frontend cookie doesn't clear itself from the previous user. Maybe it is an issue related to IIS 7.5. Never had that issue before on an Azure-hosted website or locally while developing.

**Reply by DrewBrasher** [reply]
I compared the value of the orchauth_Default cookie and it is changing when I log out and back in even when the username getting displayed is not changing.

The weirdest part to me is that I can be logged in and it showing the correct username, then someone on another computer logs in and I refresh my page and their username gets displayed on my page.

**Reply by DrewBrasher** [reply]
I'm using the default Data Protection key management. I think I will try setting it to persist to the file system and see if that helps.
https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/overview?view=aspnetcore-6.0#persistkeystofilesystem

**Reply by Skrypt** [reply]
I would try to disable the "Response Cache" feature in Orchard Core to see if it's not a server HTML cache issue. Maybe the HTML cache still holds the page rendered with the previous user. Also, make sure you don't have a Cache Tag inside your View.

**Reply by DrewBrasher** [reply]
I will try that

**Reply by DrewBrasher** [reply]
Thank you @Skrypt! I didn't see a feature called Response Cache but I disabled the one called "Dynamic Cache" and that seems to have resolved the issue. Thanks a lot for helping me with this! I've spent days trying to figure this out.

**Reply by Skrypt** [reply]
Ok, then if you have a cache tag on your page, the solution is to vary it by "user".

https://docs.orchardcore.net/en/dev/docs/reference/modules/DynamicCache/

**Reply by DrewBrasher** [reply]
This is how I was displaying the login menu:
```
{% shape "LoginMenu", alias: "alias:login-menu", cache_id: "login-menu", cache_tag: "alias:login-menu" %}
```

**Reply by Skrypt** [reply]
Cached by cache-id without a vary-by. That's why. 😄

**Reply by DrewBrasher** [reply]
I tried changing it (in Menu.liquid) to:
```
{% shape "LoginMenu", alias: "alias:login-menu", cache_id: "login-menu", cache_tag: "alias:login-menu", vary_by: "user" %}
```
but that did not work so I tried this:
```
{% cache "login-menu", vary_by: "user" %}
	{% shape "LoginMenu", alias: "alias:login-menu" %}
{% endcache %}
```
and it still did not work until I also changed this part of the Layout.liquid:
```
{% shape "menu", alias: "alias:main-menu", cache_id: "main-menu", cache_tag: "alias:main-menu" %}
```
to: 
```
{% cache "main-menu", vary_by: "user" %}
	{% shape "menu", alias: "alias:main-menu" %}
{% endcache %}
```
and now it works with Dynamic Cache enabled.

**Reply by jtkech** [reply]
Hmm as I remember, so here you use the mvc cache tag in place of the shape caching that uses the dynamic cache (which uses the distributed cache). Can you retry with only the shape tag but by also using `cache_context: "user"`
